hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -541,6 +546,10 @@ def available(self, data):
         machine = data.draw(self_strategy)
         return bool(machine.bundle(self.name))
 
+    def flatmap(self, expand):
+        self.draw_references = False
+        return super().flatmap(expand)
```

Comment:
We need to avoid mutating the strategy, because there's a good chance that the bundle is _also_ used in other places.  We can check that by adding another method to the test machine which just uses the bundle directly.

```suggestion
    def flatmap(self, expand):
        if not self.draw_references:
            return type(self)(
                self.name, consume=self.consume, draw_references=False
            ).flatmap(expand)
        return super().flatmap(expand)
```

Reply:
Is the purpose of having the extra check `if not self.draw_references:` instead of generating a new strategy each time, mainly for performance?

Reply:
Yep, primarily aimed at ensuring we hit caches where possible - the actual cost of constructing the strategy object isn't too bad but if this is an argument to something more complicated then caching that vs not might matter.